File: app/src/main/java/me/varlez/movies/movie/detail/view/MovieDetailFragment.java
package me.varlez.movies.movie.detail.view;

import android.os.Bundle;
import android.support.annotation.NonNull;
import android.support.design.widget.CollapsingToolbarLayout;
import android.support.design.widget.FloatingActionButton;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageView;
import android.widget.RelativeLayout;
import android.widget.TextView;

import com.hannesdorfmann.mosby.mvp.lce.MvpLceFragment;
import com.squareup.picasso.Picasso;

import javax.inject.Inject;

import butterknife.Bind;
import butterknife.ButterKnife;
import me.varlez.movies.MoviesApp;
import me.varlez.movies.R;
import me.varlez.movies.common.di.component.MoviesComponent;
import me.varlez.movies.common.model.Movie;
import me.varlez.movies.movie.detail.presenter.DefaultMovieDetailPresenter;
import me.varlez.movies.movie.detail.presenter.MovieDetailPresenter;
import me.varlez.movies.movie.list.view.MovieListActivity;

/**
 * A fragment representing a single Movie detail screen.
 * This fragment is either contained in a {@link MovieListActivity}
 * in two-pane mode (on tablets) or a {@link MovieDetailActivity}
 * on handsets.
 */
public class MovieDetailFragment extends MvpLceFragment<RelativeLayout, Movie, MovieDetailView, MovieDetailPresenter> implements MovieDetailView {
    /**
     * The fragment argument representing the item ID that this fragment
     * represents.
     */
    public static final String ARG_MOVIE_ID = "movieId";

    @Bind(R.id.movie_director)
    TextView directorTextView;

    @Bind(R.id.movie_plot)
    TextView plotTextView;

    @Inject
    Picasso picasso;

    private MoviesComponent moviesComponent;

    public MovieDetailFragment() {
    }

    @Override
    public void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        moviesComponent = ((MoviesApp) getActivity().getApplication()).getMoviesComponent();
        moviesComponent.inject(this);
    }

    @Override
    public View onCreateView(LayoutInflater inflater, ViewGroup container,
                             Bundle savedInstanceState) {
        return inflater.inflate(R.layout.movie_detail, container, false);
    }

    @Override
    public void onDestroyView() {
        super.onDestroyView();
        ButterKnife.unbind(this);
    }

    @Override
    public void onViewCreated(View view, Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);

        ButterKnife.bind(this, view);

        FloatingActionButton fab = (FloatingActionButton) getActivity().findViewById(R.id.fab);
        fab.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                presenter.openImdb(getActivity());
            }
        });

        if (getArguments().containsKey(ARG_MOVIE_ID)) {
            loadData(false);
        }
    }

    @Override
    public void setData(Movie data) {
        CollapsingToolbarLayout appBarLayout = (CollapsingToolbarLayout) getActivity().findViewById(R.id.toolbar_layout);
        ImageView headerImage = (ImageView) getActivity().findViewById(R.id.imageview_header);

        if (appBarLayout != null) {
            appBarLayout.setTitle(data.getTitle());
        }

        picasso.load(data.getPosterUrl()).fit().centerCrop().into(headerImage);
        plotTextView.setText(data.getPlot());
        directorTextView.setText(data.getDirector());
    }

    @Override
    public void loadData(boolean pullToRefresh) {
        presenter.details(getArguments().getString(ARG_MOVIE_ID), pullToRefresh);
    }

    @NonNull
    @Override
    public MovieDetailPresenter createPresenter() {
        return new DefaultMovieDetailPresenter(moviesComponent);
    }

    @Override
    protected String getErrorMessage(Throwable e, boolean pullToRefresh) {
        return e.getMessage();
    }
}

File: app/src/main/java/me/varlez/movies/movie/detail/view/MovieDetailFragment.kt
package me.varlez.movies.movie.detail.view

import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.RelativeLayout
import com.hannesdorfmann.mosby3.mvp.lce.MvpLceFragment
import com.squareup.picasso.Picasso
import kotlinx.android.synthetic.main.fragment_movie_detail.*
import me.varlez.movies.MoviesApp
import me.varlez.movies.R
import me.varlez.movies.common.di.component.MoviesComponent
import me.varlez.movies.common.model.Movie
import me.varlez.movies.movie.detail.presenter.DefaultMovieDetailPresenter
import me.varlez.movies.movie.detail.presenter.MovieDetailPresenter
import me.varlez.movies.movie.list.view.MovieListActivity
import javax.inject.Inject

/**
 * A fragment representing a single Movie detail screen.
 * This fragment is either contained in a [MovieListActivity]
 * in two-pane mode (on tablets) or a [MovieDetailActivity]
 * on handsets.
 */
class MovieDetailFragment : MvpLceFragment<RelativeLayout, Movie, MovieDetailView, MovieDetailPresenter>(), MovieDetailView {

    @Inject
    internal lateinit var picasso: Picasso

    private lateinit var moviesComponent: MoviesComponent

    override fun onCreate(savedInstanceState: Bundle?) {
        moviesComponent = MoviesApp.get(context).moviesComponent
        super.onCreate(savedInstanceState)
        moviesComponent.inject(this)
    }

    override fun onCreateView(inflater: LayoutInflater?, container: ViewGroup?,
                              savedInstanceState: Bundle?): View? {
        return inflater!!.inflate(R.layout.fragment_movie_detail, container, false)
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)

//        setSupportActionBar(toolbar)
//        supportActionBar?.setDisplayHomeAsUpEnabled(true)
        
        fab?.setOnClickListener { presenter.openImdb(activity) }

        if (arguments.containsKey(ARG_MOVIE_ID)) {
            loadData(false)
        }
    }

    override fun setData(data: Movie) {
        toolbar_layout?.title = data.title

        picasso.load(data.posterUrl)
                .fit()
                .centerCrop()
                .into(imageview_header)
        movie_plot_textview?.text = data.plot
        movie_director_textview?.text = data.director
    }

    override fun loadData(pullToRefresh: Boolean) {
        presenter.details(arguments.getString(ARG_MOVIE_ID), pullToRefresh)
    }

    override fun createPresenter(): MovieDetailPresenter {
        return DefaultMovieDetailPresenter(moviesComponent)
    }

    override fun getErrorMessage(e: Throwable, pullToRefresh: Boolean): String {
        return e.message!!
    }

    companion object {
        /**
         * The fragment argument representing the item ID that this fragment
         * represents.
         */
        val ARG_MOVIE_ID = "movieId"
    }
}
